### DISCUSSION TITLE
On LSPs

### DISCUSSION THREAD

**Discussion by hyperswine** [racket/rhombus#405] [General]
I wanted to start a conversation on language servers as a very useful yet under appreciated feature that has a lot more potential than people think.

Having used rust-analyzer I found that I could do most of the work within the editor context itself and hardly had to touch the terminal. Things like good test integration, function runners, static analysis, etc. But potentially even more than just that.

I think being able to effectively integrate a language's internals like rhombus with a language server could really set it apart, especially as a kind of "successor" to lisp.

**Comment by Lazerbeak12345** [comment]
I agree that LSP is a good thing. I've been using LSPs since neovim added it natively, and it has a few important features, in order of importance (for my workflow anyway)

1. If the server is fast enough, instant debugging. (Racket-langserver is not fast enough and is usually slower than cli, usually the #lang's fault. Improved with CZ)
2. If the language has a documentation standard, it can provide function documentation without leaving the editor (especially useful over ssh, which rls doesn't always support due to dependency on a graphics stack which might not be present, also rls doesn't support documentation anyway, in part due to the move away from inline documentation, though not impossible with current tooling)
3. Formatting, while I don't use it in this way, could be done by the LSP (in theory racket could integrate this with their current formatting toolchain, but the unofficial team maintaining rls  already have their work cut out for them)
4. Syntax integrations. LSP is perhaps the only cross-language protocol that is even _capable_ of describing context-specific-syntax to the degree that LOP frameworks like racket can do. (Racket has only part of the tooling set up for this, and I don't know if rls supports this)

In fact, I've held the opinion that LSP should be used as an internal API for how drracket works for years. If crafted wisely, LSP, TreeSitter, the Debug protocol and more could all be exposed, each using the #lang providers' definitions and behaviors.

**Reply by Lazerbeak12345** [reply]
RE: "fast enough:" less than 2 seconds after a change. Most LSPs can handle this.

**Comment by samth** [comment]
I'm not sure what is being suggested here. Racket has an LSP server; it's in the racket-langserver package. It relies on all the usual Racket backend work that DrRacket uses, and should work for rhombus.

I'm sure there's more work to do to improve it, and there's lots of inspiration and code that can be taken from DrRacket and racket-mode, but I don't see the connection specifically to rhombus.

**Reply by Lazerbeak12345** [reply]
The naming of a lot of racket's components might also be confusing here. RLS _should_ support rhombus as far as I'm aware (haven't tried it yet).

Perhaps a different discussion needs to be brought up between distinguishing in name rhombus (or whatever it will be called) from the toolchain/ecosystem that includes racket, its foundations and racket's sibling languages/libraries and tools.
